### PR TITLE
Rework logical operation

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3440

## What this PR does / why we need it:
Rework logical operations with cgo.

### CODE REVIEW

File: cgo/logic.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+    } else {
+        for (uint64_t i = 0; i < n; i++) {
+            rt[i] = at[i] && bt[i];
+        }
+        if (anulls != NULL) {
+            for (uint64_t i = 0; i < n; i++) {
+                if (Bitmap_Contains(anulls, i) &&  !Bitmap_Contains(bnulls, i)) {
+                    if (!bt[i]) {
```
Comment (daviszhen) [inline cgo/logic.c:57]: premote the '!bt[i]' to the first one in line 56.

Comment (daviszhen) [reply cgo/logic.c:57]: Only process the situation:
if (anulls != NULL && bnulls != NULL) {

......
}else if(anulls != NULL){
...

}else if(bnulls != NULL){

......
}

File: cgo/logic.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mo_impl.h"
+
+// Logical and operator
```
Comment (daviszhen) [inline cgo/logic.c:19]: Add the true table in the comments.

File: cgo/logic.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    bool *at = (bool *) a;
+    bool *bt = (bool *) b;
+    if ((flag & LEFT_IS_SCALAR) != 0) {
+        for (uint64_t i = 0; i < n; i++) {
+            rt[i] = at[0] || bt[i];
+        }
+        if (rnulls != NULL) {
+            if (at[0]) {
+                for (uint64_t i = 0; i < n; i++) {
+                    if (Bitmap_Contains(rnulls, i)) {
```
Comment (daviszhen) [inline cgo/logic.c:88]: remove the check.

File: cgo/logic.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                        Bitmap_Remove(rnulls, i);
+                    }
+                }
+            }
+        }
+    } else {
+        for (uint64_t i = 0; i < n; i++) {
+            rt[i] = at[i] || bt[i];
+        }
+        if (anulls != NULL) {
```
Comment (daviszhen) [inline cgo/logic.c:111]: spilt into three conditions.

**Review by daviszhen — CHANGES_REQUESTED** [review]

File: cgo/logic.c [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    bool *at = (bool *) a;
+    bool *bt = (bool *) b;
+    if ((flag & LEFT_IS_SCALAR) != 0) {
+        for (uint64_t i = 0; i < n; i++) {
+            rt[i] = at[0] && bt[i];
+        }
+        if (rnulls != NULL) {
+            if (!at[0]) {
+                for (uint64_t i = 0; i < n; i++) {
+                    if (Bitmap_Contains(rnulls, i)) {
```
Comment (fengttt) [inline cgo/logic.c:31]: Isn't this just clear all bitmap?

File: pkg/vectorize/sum/sum.go [hunk 6]
```diff
@@ -94,3 +96,21 @@ func floatSumSels[T constraints.Float](xs []T, sels []int64) T {
 	}
 	return res
 }
+
+/*
```
Comment (fengttt) [inline pkg/vectorize/sum/sum.go:100]: Please do not add these dead code.   If you really want to vectorize sum, do it completely in another PR.

Comment (qingxinhome) [reply pkg/vectorize/sum/sum.go:100]: OK, sorry, I submitted it carelessly

File: cgo/logic.c (hunk 5 above)
Comment (qingxinhome) [reply cgo/logic.c:31]: Yes, in this case, all bitmap elements will be reset to 0

File: cgo/logic.c (hunk 4 above)
Comment (qingxinhome) [reply cgo/logic.c:111]: OK， no problem
